### PR TITLE
Rustc 178 fixes v1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
None

Describe changes:
- Fixes tests with debug and rust 1.78

### CODE REVIEW

File: src/app-layer-detect-proto.c [hunk 1]
```diff
@@ -1431,7 +1431,7 @@ AppProto AppLayerProtoDetectGetProto(AppLayerProtoDetectThreadCtx *tctx, Flow *f
         }
     }
 
-    if (!FLOW_IS_PP_DONE(f, flags)) {
+    if (!FLOW_IS_PP_DONE(f, flags) && buf != NULL) {
```
Comment (victorjulien) [inline src/app-layer-detect-proto.c:1434]: would this break probing parsers that only use the port?

File: src/detect-icode.c [hunk 2]
```diff
@@ -311,7 +311,6 @@ static int DetectICodeParseTest08(void)
     DetectU8Data *icd = DetectU8Parse("> 8 <> 20");
     FAIL_IF_NOT_NULL(icd);
 
-    DetectICodeFree(NULL, icd);
```
Comment (victorjulien) [inline src/detect-icode.c:314]: should probably annotate the func with nonnull attribute for that ptr

Comment (catenacyber) [reply src/detect-icode.c:314]: Is it done in some other function ? (`git grep nonnull` does not show anything like that)

File: rust/src/applayer.rs [hunk 3]
```diff
@@ -58,6 +58,9 @@ impl StreamSlice {
         self.input_len
     }
     pub fn as_slice(&self) -> &[u8] {
+        if self.input.is_null() && self.input_len == 0 {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/applayer.rs:61]: Not sure : you do not want to hide gaps if you claim to handle them...

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20527

File: src/detect-icode.c (hunk 2 above)
Comment (victorjulien) [reply src/detect-icode.c:314]: iirc it is `__attribute__((nonnull)))`

File: rust/src/applayer.rs [hunk 4]
```diff
@@ -386,6 +386,17 @@ macro_rules! build_slice {
     ($buf:ident, $len:expr) => ( std::slice::from_raw_parts($buf, $len) );
 }
 
+#[macro_export]
+macro_rules! build_slice_null {
```
Comment (victorjulien) [inline rust/src/applayer.rs:390]: seems to belong in it's own commit

Also, should this be used in the `as_slice` thing as well?

Comment (catenacyber) [reply rust/src/applayer.rs:390]: It is in its own commit. What do you mean ?

> Also, should this be used in the as_slice thing as well?

I think I tried and did not get it to compile with magic macro...
But then, we would also have to refactor all usages of from_raw_parts to use either macro...

File: src/app-layer-detect-proto.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-detect-proto.c:1434]: I think this is something we told ppl that is supported, so best to keep it supported

File: rust/src/applayer.rs (hunk 4 above)
Comment (victorjulien) [reply rust/src/applayer.rs:390]: `build_slice_null` addition in its own commit, explaining it, instead of making it part of some http2 commit

**Comment by victorjulien** [comment]
We'll need a ticket too, as we'll need to backport things to both 6 and 7

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/11037

File: rust/src/applayer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/applayer.rs:61]: That makes sense to me

File: src/app-layer-detect-proto.c (hunk 1 above)
Comment (jasonish) [reply src/app-layer-detect-proto.c:1434]: > I think this is something we told ppl that is supported, so best to keep it supported

This would apply to those with custom parsers?
